save-backend/src/main/kotlin/org/cqfn/save/backend/configs/WebConfiguration.kt

```diff
@@ -43,11 +71,38 @@ class WebConfiguration(
         @Value("classpath:/static/error.html") errorPage: Resource,
     ) = router {
         GET("/") {
-            ok().header("Content-Type", "text/html; charset=utf8").bodyValue(indexPage)
+            ok().header("Content-Type", "text/html; charset=utf8")
+                .cacheControl(shortExpirationTime) { cachePublic() }
+                .lastModified(buildProperties.time)
+                .bodyValue(indexPage)
         }
 
         GET("/error") {
-            ok().header("Content-Type", "text/html; charset=utf8").bodyValue(errorPage)
+            ok().header("Content-Type", "text/html; charset=utf8")
+                .cacheControl(shortExpirationTime) { cachePublic() }
+                .lastModified(buildProperties.time)
+                .bodyValue(errorPage)
         }
     }
+
+    private fun RouterFunctionDsl.cacheableFsResource(
+        pattern: String,
+        basePath: String,
+    ) = GET(pattern) { request ->
+        val resourcePath = request.pathVariable("resourcePath")
+        val resource = FileSystemResource("$basePath/$resourcePath")
+        ok().cacheControl(longExpirationTime) { cachePublic() }
+            .lastModified(resource.lastModified().toInstant())
+            .bodyValue(resource)
+    }
+
+    private fun ServerResponse.BodyBuilder.cacheControl(duration: kotlin.time.Duration,
+                                                        cacheControlCustomizer: CacheControl.() -> CacheControl = { this },
+    ): ServerResponse.BodyBuilder =
+            cacheControl(CacheControl.maxAge(duration.toJavaDuration()).run(cacheControlCustomizer))
+
+    companion object {
+        private val shortExpirationTime = 10.minutes
```

Comment:
Could you please describe your thoughts and observations about how it actually worked on practice?

 * After building the jar, if the time which is passed, less, than 10 minutes, than we won't make requests for checking the content and for uploading resources at all?
* If the elapsed time is more than 10 min, then we will make request, to check if the content (or some checksum) is changed, and only in this case will upload resources one more time?
* If content is the same, even, if it pass a lot of time, we keep our cache as is?

Reply:
Also it will be good, if we won't download them always, if content is the same, but the elapsed time is large than 10 min
But this is not so critical, since before this moment we had no cache at all

Reply:
There is an RFC about caching: https://datatracker.ietf.org/doc/html/rfc7234#section-5.2, also docs on MDN: https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching#freshness.
Correct combination of `Last-Modified` in response and `If-Modified-Since` in request should make browsers cache content, while periodically querying the server and getting `304 Not Changed`. After we deploy a new version of save-cloud, `Last-Modified` changes and browser will (after some time) revalidate its cache. As I understand it, with `max-age=600` the browser must validate resource freshness each 10 minutes, after getting `304 Not Modified` it will reset resource age and serve it from cache for another 10 minutes.
That being said, I've noticed that Chrome caches content with only `Cache-Control`, while Firefox needs `Last-Modified` or `ETag` too.

There is [another technique](https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching#revved_resources) to ensure freshness of important resources (HTML, JS, CSS) by adding content hash or version directly to the filename (webpack supports this functionality out of the box). This way new files will have a different URL and will be downloaded unconditionally. We may add this in this PR or in the future too.

Reply:
Great, thanks for detailed answer

About `revving` - good technique, I think, that we can play with it in future